runtime/vm/UpcallVMHelpers.cpp

```diff
@@ -489,6 +455,67 @@ convertUpcallReturnValue(J9UpcallMetaData *data, U_8 returnType, U_64 *returnSto
 	}
 }
 
+/**
+ * @brief Allocate memory related objects for the struct/pointer arguments
+ * to store them to the java array before placing them on the java stack
+ * for upcall.
+ *
+ * @param data a pointer to J9UpcallMetaData
+ * @param currentThread a pointer to J9VMThread
+ * @return true if the storing operation is completed; otherwise return false;
+ */
+static bool
+storeMemArgObjectsToJavaArray(J9UpcallMetaData *data, J9VMThread *currentThread)
+{
+	J9UpcallNativeSignature *nativeSig = data->nativeFuncSignature;
+	J9UpcallSigType *sigArray = nativeSig->sigArray;
+	I_32 paramCount = (I_32)(nativeSig->numSigs - 1); /* The last element is for the return type. */
+	bool result = true;
+
+	for (I_32 argIndex = 0; argIndex < paramCount; argIndex++) {
+		U_8 argSigType = sigArray[argIndex].type & J9_FFI_UPCALL_SIG_TYPE_MASK;
+		j9object_t memArgObject = NULL;
+		j9object_t nativeArgArray = NULL;
+
+		if ((J9_FFI_UPCALL_SIG_TYPE_POINTER == argSigType)
+			|| (J9_FFI_UPCALL_SIG_TYPE_STRUCT == argSigType)
+		) {
+			/* Temporarily add an explicit call to the GC after the object allocations in this code,
+			 * to prove that the stack walker is operating correctly.
+			 */
+			currentThread->javaVM->memoryManagerFunctions->j9gc_modron_local_collect(currentThread);
```

Comment:
Maybe add a global GC call here too to move as many objects as possible.

Reply:
Added a global GC call as suggested above.